src/repository.c

```diff
@@ -1066,6 +1066,74 @@ void git_repository_set_config(git_repository *repo, git_config *config)
 	set_config(repo, config);
 }
 
+static int load_grafts(git_odb *odb, git_repository *repo)
+{
+	git_buf graft_path = GIT_BUF_INIT;
+	git_buf contents = GIT_BUF_INIT;
+	const char *line_start;
+	const char *line_end;
+	int line_num = 0;
+	int error, updated;
+	git_array_oid_t parents = GIT_ARRAY_INIT;
+
+	if ((error = git_repository_item_path(&graft_path, repo, GIT_REPOSITORY_ITEM_INFO)) < 0)
+		return error;
+
+	if (git_buf_joinpath(&graft_path, graft_path.ptr, "grafts")) {
+		git_buf_free(&graft_path);
```

Comment:
Maybe have a common `goto cleanup` again? It's becoming hard to grasp what needs cleaning at which part of the code with those interspersed `free`s and `return`s.